apps/worker/src/monitor/tx.status.monitor.service.ts

```diff
@@ -0,0 +1,130 @@
+import { SECONDS_PER_BLOCK } from '../../../../libs/common/src/constants';
+
+@Injectable()
+@Processor(QueueConstants.TRANSACTION_RECEIPT_QUEUE_NAME, {
+  concurrency: 2,
+})
+export class TxStatusMonitoringService extends WorkerHost implements OnApplicationBootstrap, OnModuleDestroy {
+  private logger: Logger;
+
+  constructor(
+    @InjectRedis() private cacheManager: Redis,
+    @InjectQueue(QueueConstants.TRANSACTION_RECEIPT_QUEUE_NAME) private txReceiptQueue,
+    @InjectQueue(QueueConstants.PUBLISH_QUEUE_NAME) private publishQueue: Queue,
+    private blockchainService: BlockchainService,
+    private configService: ConfigService,
+    private eventEmitter: EventEmitter2,
+  ) {
+    super();
+    this.logger = new Logger(this.constructor.name);
+  }
+
+  public async onApplicationBootstrap() {
+    this.logger.debug('Starting publishing service');
+  }
+
+  public onModuleDestroy() {
+    try {
+      this.logger.debug('Shutting down publishing service');
+    } catch (e) {
+      // 🐂 //
+    }
+  }
+
+  async process(job: Job<ITxMonitorJob, any, string>): Promise<any> {
+    this.logger.log(`Monitoring job ${job.id} of type ${job.name}`);
+    try {
+      const numberBlocksToParse = 100n;
+      const txCapacityEpoch = job.data.epoch;
+      const previousKnownBlockNumber = (await this.blockchainService.getBlock(job.data.lastFinalizedBlockHash)).block.header.number.toBigInt();
+      const currentFinalizedBlockNumber = await this.blockchainService.getLatestFinalizedBlockNumber();
+      const blockList: bigint[] = [];
+      for (let i = previousKnownBlockNumber; i <= currentFinalizedBlockNumber && i < previousKnownBlockNumber + numberBlocksToParse; i += 1n) {
+        blockList.push(i);
+      }
+      const txBlockHash = await this.crawlBlockList(job.data.txHash, txCapacityEpoch, blockList);
+
+      if (txBlockHash) {
+        this.logger.verbose(`Successfully completed job ${job.id}`);
+        return { success: true };
+      }
+      return { success: false };
+    } catch (e) {
+      this.logger.error(`Job ${job.id} failed (attempts=${job.attemptsMade}) with error: ${e}`);
+      throw e;
+    } finally {
+      // do some stuff
+    }
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  @OnWorkerEvent('completed')
+  onCompleted() {
+    // do some stuff
+  }
+
+  private async crawlBlockList(txHash: Hash, epoch: string, blockList: bigint[]): Promise<BlockHash | undefined> {
+    // Use Promise.all to await all filter operations before returning undefined
+    await Promise.all(
+      blockList.map(async (blockNumber) => {
+        const blockHash = await this.blockchainService.getBlockHash(blockNumber);
+        const block = await this.blockchainService.getBlock(blockHash);
+        const txInfo = block.block.extrinsics.filter((extrinsic) => extrinsic.hash.toString() === txHash.toString());
+        this.logger.debug(`Extrinsics: ${block.block.extrinsics[0]}`);
+
+        if (txInfo.length > 0) {
```

Comment:
Since we only care if the extrinsic is in the block (and should only appear once, anyway), `.find()` is probably more efficient than `.filter()`. Also why are you logging `extrinsics[0]`? I imagine you either want the whole array, or just the found item...

```suggestion
        const txInfo = block.block.extrinsics.find((extrinsic) => extrinsic.hash.toString() === txHash.toString());
        this.logger.debug(`Extrinsics: ${block.block.extrinsics[0]}`);

        if (txInfo !== undefined) {
```